nixos/modules/services/databases/kanidm.nix

```diff
@@ -1,112 +1,335 @@
+    PrivateNetwork = true;
+    PrivateTmp = true;
+    PrivateUsers = true;
+    ProcSubset = "pid";
+    ProtectClock = true;
+    ProtectHome = true;
+    ProtectHostname = true;
+    # Would re-mount paths ignored by temporary root
+    #ProtectSystem = "strict";
+    ProtectControlGroups = true;
+    ProtectKernelLogs = true;
+    ProtectKernelModules = true;
+    ProtectKernelTunables = true;
+    ProtectProc = "invisible";
+    RestrictAddressFamilies = [ ];
+    RestrictNamespaces = true;
+    RestrictRealtime = true;
+    RestrictSUIDSGID = true;
+    SystemCallArchitectures = "native";
+    SystemCallFilter = [ "@system-service" "~@privileged @resources @setuid @keyring" ];
+    # Does not work well with the temporary root
+    #UMask = "0066";
+  };
+
 in
 {
+  options = {
+    services.kanidm = {
+      enable = lib.mkEnableOption "the kanidm server";
+      enablePam = lib.mkEnableOption "the pam and nsswitch integration of kanidm";
 
-  options.services.kanidm = {
-    enable = lib.mkEnableOption "kanidm service";
-
-    ensureDomainName = lib.mkOption {
-      description = ''
-        The domain_name that Kanidm manages. It must be a descendant or equal to the domain
-        specified in settings.origin. It is always set when the server starts up.
-        While in theory it could be changed later, you should carefully consider these
-        warnings:
-        <link xlink:href="https://kanidm.github.io/kanidm/administrivia.html#rename-the-domain"/>
-      '';
-      example = "example.org";
-      type = lib.types.str;
-    };
+      package = lib.mkOption {
+        type = lib.types.package;
+        default = pkgs.kanidm;
+        description = "Which kanidm package to use.";
+      };
 
-    settings = lib.mkOption {
-      type = lib.types.submodule {
+      ensureDomainName = lib.mkOption {
+        description = ''
+          The domain_name that Kanidm manages. It must be a descendant or equal to the domain
+          specified in settings.origin. It is always set when the server starts up.
+          While in theory it could be changed later, you should carefully consider these
+          warnings:
+          <link xlink:href="https://kanidm.github.io/kanidm/administrivia.html#rename-the-domain"/>
+        '';
+        example = "example.org";
+        type = lib.types.str;
+      };
 
-        freeformType = settingsFormat.type;
+      serverSettings = lib.mkOption {
+        type = lib.types.submodule {
+          freeformType = settingsFormat.type;
 
-        options.bindaddress = lib.mkOption {
-          description = "Address/port combination the webserver binds to.";
-          example = "[::1]:8443";
-          type = lib.types.str;
-        };
-        options.origin = lib.mkOption {
-          description = "The origin of your kanidm instance. Must have https as protocol.";
-          example = "https://idm.example.org";
-          type = lib.types.str;
+          options.bindaddress = lib.mkOption {
```

Comment:
Instead of nesting every element we IMO should pull options out of here and and instead indent this block once more.